jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/CompositeNodeInstance.java

```diff
@@ -122,9 +122,13 @@ public void internalTrigger(final org.kie.api.runtime.process.NodeInstance from,
 	        List<Connection> connections = nodeAndType.getNode().getIncomingConnections(nodeAndType.getType());
 	        for (Iterator<Connection> iterator = connections.iterator(); iterator.hasNext(); ) {
 	            Connection connection = iterator.next();
+                long fromNodeId = from != null ? from.getNodeId() : -1;
+                if (from != null && from.getNode() instanceof AsyncEventNode) {
+                    fromNodeId = ((AsyncEventNode) from.getNode()).getPreviousNode(type).getId();
+                }
 	            if ((connection.getFrom() instanceof CompositeNode.CompositeNodeStart) &&
-	            		(from == null ||
-	    				((CompositeNode.CompositeNodeStart) connection.getFrom()).getInNode().getId() == from.getNodeId())) {
+                    (from == null ||
+                     ((CompositeNode.CompositeNodeStart) connection.getFrom()).getInNode().getId() == fromNodeId)) {
```

Comment:
if the exclude based on metadata will work, then I think all this code change won't be needed

Reply:
I cannot remove this part. the from is performed before unwrapping so this step is needed. The unwrap is not needed anymore.

Reply:
how about finding this logic in getFrom so it actually inspects if the from node is async and if so will use its actualNode, would that work?

Reply:
@mswiderski 
I am not sure how this should work. In this particular case when you are operating with async = false you get

ActionNode (connection.getFrom.getInNode) -> ActionNode (from node) -> ForEach (next)

with async = true you get
ActionNode (connection.getFrom.getInNode) -> AsyncNode (from node, ActualNode = forEach)

so when you are getting the from.getNodeId you are getting the foreach node and not the action node id (the previous node from the actual node) causing a mismatch and not able to go through the proper control flow. changin any call in  asyncnodeinstance (getNodeId, getNode) is not working because it causes collateral problems (unsolvable from my point of view)

changing getFrom connection won't work either as getFrom node does not hold the wrap information.

The call getFrom is working in both cases as it is receiving the proper node. If you get de NodeInstance (an async node) the workflow won't advance at that stage properly. that is the reason we need to avoid wrapping again in a AsyncNodeInstance (we would be triggering several times the async node instead of the actual node -> for each)

I don't see a way to work this same way or sink the logic anywhere in the code

Reply:
In that case I can only say I trust your investigation and apparently you're right and this can't be moved anyway. So I think we are good to go with it. @MarianMacik can you have a look as well?

Reply:
@MarianMacik it would be great if this makes to the cut off this friday. Let's see if we can make it.

Reply:
So here we are matching connection.getFrom.getInNode (the node which is before CompositeStartNode) with the from node which is argument of internalTrigger method, right? If there is a match, we can trigger and continue... In case of sync it works ok since the from is a node which is before foreach node/composite node. But in case of async the from node is already the AsyncNode which wraps the foreach/composite node we are trying to trigger? In that case I think that we should properly pass the real from node (so the same node as with sync - the one before the foreach node) into the method - if it is possible. @mswiderski isn't an issue in that part - that we are passing the incorrect from node which is not the from node but rather a wrapper of the current node?

Reply:
@MarianMacik  Just to point out that with an async execution you are changing the bpmn adding a new node instance in between ActionNode and ForEachNode ... AsynEventNode

Reply:
well from what you wrote and from the code I understood that engine wraps it, not inserts it, doesn't it?

Reply:
Depends how you look at it...  the usual stack trace 
StartNode -> Action Node -> ForEach -> Action Node.... in this case... so the  from that is pass to the for each is from action node
This changes from async
Start Node -> AsyncNode (Action Node) ->  AsyncNode (For Each) -> AsyncNode (Action Node)

from node is the async node... not where it does come from. That is the whole problem. If you try to chain like normal flow you will get an infinite loop because asyncnode instance.
The algorithm of execution does not work the same way.